### PR TITLE
Feature/blackpillv2: add alternative pinout

### PR DESCRIPTION
**Pull request by lenvm**
## Detailed description
This pull request
- implements function `PINOUT_SWITCH()` to switch between pinouts. In total 3 alternative pinouts can be defined.
- adds one alternative pinout for the blackpillv2 platform. The pinout description is found in the README.md of the blackpillv2 platform, modified in this pull request. This pinout uses the pins near the bottom of the board. The alternative pinout is visualized in figure 1.
- adds the macros to facilitate using either USART1 or USART2 as USBUSART.

The implementation of a function to select alternative pinouts was mentioned in [this issue comment](https://github.com/blackmagic-debug/blackmagic/issues/1418#issuecomment-1507492647).

230974447-ec7b25cb-ee38-4424-a54b-d596888b1aa3
Figure 1: alternative pinout 1

## Your checklist for this pull request
* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`).
* [x] It builds for hardware blackpillv2 (`make PROBE_HOST=blackpillv2`)
* [x] It builds for hardware blackpillv2 with alternative pinout 1 (`make PROBE_HOST=blackpillv2 ALTERNATIVE_PINOUT=1`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

File: src/platforms/blackpillv2/README.md [hunk 1]
```diff
@@ -4,22 +4,22 @@ Allows the use of a [BlackPill v2](https://github.com/WeActStudio/WeActStudio.Mi
...
+| nRST            | PA5    | PB4                  | JTAG/SWD  |
+| TRST (optional) | PA6    | PB3                  | JTAG/SWD  |
+| UART TX         | PA2    | PA2                  | USB USART |
+| UART RX         | PA3    | PA3                  | USB USART |
+| Power pin       | PA1    | PB9                  | Power     |
+| LED idle run    | PC13   | PC13                 | LED       |
+| LED error       | PC14   | PC14                 | LED       |
+| LED bootloader  | PC15   | PC15                 | LED       |
+| LED UART        | PA4    | PA1                  | LED       |
+| User button KEY | PA0    | PA0                  |           |
```
Comment (dragonmux) [inline src/platforms/blackpillv2/README.md:24]: This extra newline should be removed

File: src/platforms/blackpillv2/README.md [hunk 2]
```diff
@@ -30,6 +30,15 @@ make clean
...
 
+or
+
+```sh
+cd blackmagic
+make clean
+make PROBE_HOST=blackpillv2 ALTERNATIVE_PINOUT=1
+```
+
+
```
Comment (dragonmux) [inline src/platforms/blackpillv2/README.md:41]: This extra newline should be removed

File: src/platforms/blackpillv2/Makefile.inc [hunk 3]
```diff
@@ -27,6 +27,16 @@ endif
 
 VPATH += platforms/stm32
 
+ifeq ($(ALTERNATIVE_PINOUT) <= 0, 1)
+CFLAGS += -DALTERNATIVE_PINOUT=0
+else ifeq ($(ALTERNATIVE_PINOUT), 1)
+CFLAGS += -DALTERNATIVE_PINOUT=1
+else ifeq ($(ALTERNATIVE_PINOUT), 2)
```
Comment (dragonmux) [inline src/platforms/blackpillv2/Makefile.inc:34]: Given there's only one alternative pinout currently defined, and you're mapping `$(ALTERNATIVE_PINOUT)` to `CFLAGS`' `-DALTERNATIVE_PINOUT=` values 1:1, it seems to us to make sense to pass the value through directly as `CFLAGS += -DALTERNATIVE_PINOUT=$(ALTERNATIVE_PINOUT)`

File: src/platforms/blackpillv2/platform.h [hunk 4]
```diff
@@ -29,81 +29,155 @@
 
 #define PLATFORM_HAS_TRACESWO
 #define PLATFORM_IDENT "(BlackPillV2) "
+
+/* Error handling for ALTERNATIVE_PINOUT
+ * If ALTERNATIVE_PINOUT has a value >= 4 (undefined), or <= 0, an error is thrown.
+ */
+#ifdef ALTERNATIVE_PINOUT
+#if ALTERNATIVE_PINOUT <= 0
```
Comment (dragonmux) [inline src/platforms/blackpillv2/platform.h:37]: please add `|| ALTERNATIVE_PINOUT > 3`, and drop the `#else` + `#warning` branch below - it doesn't make sense to check this condition twice in two different ways.

File: src/platforms/blackpillv2/platform.h [hunk 5]
```diff
@@ -29,81 +29,155 @@
...
 #define LED_BOOTLOADER GPIO15
 
 #define LED_PORT_UART GPIOA
-#define LED_UART      GPIO4
+#define LED_UART      PINOUT_SWITCH(GPIO4, GPIO1)
 
-/* for USART2, DMA1 is selected from https://www.st.com/resource/en/reference_manual/dm00119316-stm32f411xc-e-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf, page 170, table 27.
- * This table defines USART2_TX as stream 6, channel 4, and USART2_RX as stream 5, channel 4.
+/* USART2 with PA2 and PA3 is selected as USBUSART. Alternatively USART1 with PB6 and PB7 can be used.
  */
```
Comment (dragonmux) [inline src/platforms/blackpillv2/platform.h:118]: This should be moved onto the line above

**Review by dragonmux — CHANGES_REQUESTED** [review]
This does complicate the platform and we do have some code changes we've noted during review, but given this solves a problem with the platform for you we're happy to merge this once the review items have been addressed.

File: src/platforms/blackpillv2/README.md (hunk 1 above)
Comment (lenvm) [reply src/platforms/blackpillv2/README.md]: removed

File: src/platforms/blackpillv2/Makefile.inc (hunk 3 above)
Comment (lenvm) [reply src/platforms/blackpillv2/Makefile.inc]: Updated in the way you suggested.

I have also added `ifndef ALTERNATIVE_PINOUT` such that the code can still be built without passing the `ALTERNATIVE_PINOUT` to make.
This was already possible using the implementation you commented on, but was not possible without adding the `ifndef` line.

File: src/platforms/blackpillv2/platform.h (hunk 4 above)
Comment (lenvm) [reply src/platforms/blackpillv2/platform.h]: Implemented. Thanks!

File: src/platforms/blackpillv2/platform.h (hunk 5 above)
Comment (lenvm) [reply src/platforms/blackpillv2/platform.h]: Implemented 👍

**Review by dragonmux — APPROVED** [review]
LGTM, merging. Thank you for the contribution!

**Pull request merged**